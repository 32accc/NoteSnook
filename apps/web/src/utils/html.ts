/*
This file is part of the Notesnook project (https://notesnook.com/)

Copyright (C) 2023 Streetwriters (Private) Limited

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

export function h(
  tag: keyof HTMLElementTagNameMap | "text",
  children: (HTMLElement | string)[] = [],
  attr: Record<string, string | undefined> = {}
) {
  const element = document.createElement(tag);
  element.append(
    ...children.map((v) =>
      typeof v === "string" ? document.createTextNode(v) : v
    )
  );
  for (const key in attr) {
    const value = attr[key];
    if (value) element.setAttribute(key, value);
  }
  return element;
}

export function text(text: string) {
  return document.createTextNode(text);
}
