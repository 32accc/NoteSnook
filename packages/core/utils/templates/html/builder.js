/*
This file is part of the Notesnook project (https://notesnook.com/)

Copyright (C) 2023 Streetwriters (Private) Limited

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

import { parseHTML } from "../../html-parser";
import HTMLTemplate from "./template";

const LANGUAGE_REGEX = /(?:^|\s)lang(?:uage)?-([\w-]+)(?=\s|$)/i;

async function buildHTML(templateData) {
  return HTMLTemplate(await preprocessHTML(templateData));
}

async function preprocessHTML(templateData) {
  const { content } = templateData;
  const doc = parseHTML(
    content.replaceAll(/<p(.+?)><\/p>/gm, "<p$1><br/></p>")
  );

  const mathBlocks = doc.querySelectorAll(".math-block.math-node");
  const mathInlines = doc.querySelectorAll(".math-inline.math-node");

  if (mathBlocks.length || mathInlines.length) {
    const { default: katex } = require("katex");
    require("katex/contrib/mhchem");

    for (const mathBlock of mathBlocks) {
      const text = mathBlock.textContent;
      mathBlock.innerHTML = katex.renderToString(text, {
        displayMode: true,
        output: "mathml"
      });
    }

    for (const mathInline of mathInlines) {
      const text = mathInline.textContent;
      mathInline.innerHTML = katex.renderToString(text, { output: "mathml" });
    }
    templateData.hasMathBlocks = true;
  }

  const codeblocks = doc.querySelectorAll("pre > code");
  if (codeblocks.length) {
    const { default: prismjs } = require("prismjs");
    prismjs.register = () => {};
    for (const codeblock of codeblocks) {
      const language = LANGUAGE_REGEX.exec(
        codeblock.parentElement.className
      )?.[1];
      if (!language) continue;
      const {
        default: grammar
      } = require(`../../../../editor/languages/${language}.js`);

      grammar(prismjs);

      codeblock.innerHTML = prismjs.highlight(
        codeblock.textContent,
        prismjs.languages[language],
        language
      );
    }
    templateData.hasCodeblock = true;
  }

  templateData.content = doc.body.innerHTML;
  return templateData;
}

export default { buildHTML };
