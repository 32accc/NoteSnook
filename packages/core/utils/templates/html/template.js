/*
This file is part of the Notesnook project (https://notesnook.com/)

Copyright (C) 2022 Streetwriters (Private) Limited

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

const template = `<!DOCTYPE html>
<html lang="en-US">
  <head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=Edge,chrome=1" />
    <meta
      name="description"
      content="{{headline}}"
    />
    <title>{{title}} - Notesnook</title>
    <meta name="created-on" content="{{createdOn}}" />
    <meta name="last-edited-on" content="{{editedOn}}" />
    <meta name="tags" content="{{tags}}" />
  </head>
  <body>
    <h1>{{title}}</h1>
    {{content}}
  </body>
</html>
`;
export default template;
