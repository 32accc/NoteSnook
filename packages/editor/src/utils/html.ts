/*
This file is part of the Notesnook project (https://notesnook.com/)

Copyright (C) 2023 Streetwriters (Private) Limited

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

import { encodeNonAsciiHTML } from "entities";

export function convertBrToParagraph(html: string) {
  const doc = new DOMParser().parseFromString(html, "text/html");
  for (const br of doc.querySelectorAll("br")) {
    let paragraph = br.closest("p");

    // if no paragraph is found over the br, we add one.
    if (!paragraph && br.parentElement) {
      const parent = br.parentElement;
      const p = document.createElement("p");
      p.append(...parent.childNodes);
      parent.append(p);
      paragraph = p;
    }

    if (paragraph && paragraph.childNodes.length === 1) continue;
    if (paragraph) {
      splitOn(paragraph, br);
      const children = Array.from(paragraph.childNodes.values());
      const newParagraph = doc.createElement("p");
      newParagraph.dataset.spacing = "single";
      newParagraph.append(...children.slice(children.indexOf(br) + 1));
      paragraph.insertAdjacentElement("afterend", newParagraph);
      br.remove();
    }
  }
  return doc;
}

function splitOn(bound: Element, cutElement: Element) {
  let grandparent: ParentNode | null = null;
  for (
    let parent = cutElement.parentNode;
    bound != parent;
    parent = grandparent
  ) {
    if (parent) {
      const right = parent.cloneNode(false);
      while (cutElement.nextSibling) right.appendChild(cutElement.nextSibling);
      grandparent = parent.parentNode;
      grandparent?.insertBefore(right, parent.nextSibling);
      grandparent?.insertBefore(cutElement, right);
    }
  }
}

export function convertTextToHTML(src: string) {
  return src
    .split(/[\r\n]/)
    .map((line) =>
      line
        ? `<p data-spacing="single">${encodeLine(line)}</p>`
        : `<p data-spacing="single"></p>`
    )
    .join("");
}

function encodeLine(line: string) {
  line = encodeNonAsciiHTML(line);
  line = line.replace(/(^ +)|( {2,})/g, (sub, ...args) => {
    const [starting, inline] = args;
    if (starting) return "&nbsp;".repeat(starting.length);
    if (inline) return "&nbsp;".repeat(inline.length);
    return sub;
  });
  return line;
}
