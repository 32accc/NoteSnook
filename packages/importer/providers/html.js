const templates = require("../utils/template");
const extension = require("../utils/extension");

/**
 *
 * @param  {Array<{data:string | Buffer,createdAt:number,modifiedAt:number,fileName:string}>} files
 * @returns array of notes
 */
function convert(files) {
  let notes = [];
  for (var file of files) {
    if (typeof file.data !== "string") {
      file.data = file.data.toString()
    }
    if (extension.get(file.fileName) === "html") {
      let note = templates.note();
      note.content.data = file.data
      note.dateCreated = file.createdAt;
      note.dateEdited = file.modifiedAt;
      note.title = file.fileName.split(".")[0] || file.data.slice(0, 40);
      notes.push(note);
    } else {
      console.warn(`${file.fileName} is not a html file, skipping.`);
    }
  }
  return notes;
}

module.exports = {
  convert
};
