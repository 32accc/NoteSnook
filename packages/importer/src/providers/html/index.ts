import { ContentType, Note } from "../../models/note";
import { File } from "../../utils/file";
import {
  IFileProvider,
  iterate,
  ProviderResult,
  ProviderSettings,
} from "../provider";
import { parse } from "node-html-parser";
import { path } from "../../utils/path";

export class HTML implements IFileProvider {
  type: "file" = "file";
  public supportedExtensions = [".html"];
  public validExtensions = [...this.supportedExtensions];
  public version = "1.0.0";
  public name = "HTML";

  async process(
    files: File[],
    settings: ProviderSettings
  ): Promise<ProviderResult> {
    return iterate(this, files, (file, notes) => {
      const data = file.text;
      const document = parse(data);

      const title =
        document.querySelector("title")?.textContent ||
        document.querySelector("h1,h2")?.textContent;
      let note: Note = {
        title: title || path.basename(file.name),
        dateCreated: file.createdAt,
        dateEdited: file.modifiedAt,
        content: {
          type: ContentType.HTML,
          data: document.querySelector("body")?.innerHTML || data,
        },
      };
      notes.push(note);

      return Promise.resolve(true);
    });
  }
}
