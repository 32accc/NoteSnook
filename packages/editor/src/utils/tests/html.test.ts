/*
This file is part of the Notesnook project (https://notesnook.com/)

Copyright (C) 2023 Streetwriters (Private) Limited

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

import { test } from "vitest";
import { convertBrToParagraph } from "../html";

const cases = [
  [`<p>line 1<br>line 2</p>`],
  [`<p>line <em>1<br>line</em> 2</p>`],
  [`<p>line <span><em>1<br>line</em></span> 2</p>`],
  [`<p>line <span><em>1<br data-some="hello">line</em></span> 2</p>`],
  [`<p><br/></p>`],
  [
    `
  <html><body>
<!--StartFragment-->A troll, they call me, but I have no wish<br>
to be associated with those dolls<br>
<br>
We lack religion, purpose, politics,<br>
and yet, we somehow manage to get by.<br>
</body>
</html>`
  ],
  [
    `<html><body>
    <!--StartFragment--><p dir="auto">When I try to paste something (e.g. email content) to a note, the styling is kept, which is good, but the newlines are removed.<br>
    Also when I share the selection to Notesnook via the share functionality from Android, I have the same issue.</p>
    <hr>
    <p dir="auto"><strong>Device information:</strong><br>
    App version: 2.3.0<br>
    Platform: android<br>
    Model: OnePlus-CPH2409-31<br>
    Pro: true<br>
    Logged in: yes</p><!--EndFragment-->
    </body>
    </html>`
  ],
  [
    `<html><body>
    <!--StartFragment--><span class="css-901oao css-16my406 r-poiln3 r-bcqeeo r-qvutc0">Why switch from Gmail? 

    Not sacrificing features for more privacy, prefer using one app, in many public groups and channels (Telegram)
    
    LibreOffice Slow &amp; buggy
    
    Switched to Brave for the better Android app, more private out of the box &amp; unsure if uBlock Origin closes gap</span><!--EndFragment-->
    </body>
    </html>`
  ]
];

for (const testCase of cases) {
  const [html, expected] = testCase;
  test(`convert br tags to paragraphs (${testCase})`, (t) => {
    t.expect(
      convertBrToParagraph(html).body.innerHTML.trim()
    ).toMatchSnapshot();
  });
}
