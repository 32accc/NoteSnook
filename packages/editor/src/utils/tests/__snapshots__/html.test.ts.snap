// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`convert br tags to paragraphs (
  <html><body>
<!--StartFragment-->A troll, they call me, but I have no wish<br>
to be associated with those dolls<br>
<br>
We lack religion, purpose, politics,<br>
and yet, we somehow manage to get by.<br>
</body>
</html>) 1`] = `"<p><!--StartFragment-->A troll, they call me, but I have no wish</p><p data-spacing=\\"single\\">to be associated with those dolls</p><p data-spacing=\\"single\\"></p><p data-spacing=\\"single\\">We lack religion, purpose, politics,</p><p data-spacing=\\"single\\">and yet, we somehow manage to get by.</p><p data-spacing=\\"single\\"></p><p data-spacing=\\"single\\">  </p>"`;

exports[`convert br tags to paragraphs (<html><body>
    <!--StartFragment--><p dir="auto">When I try to paste something (e.g. email content) to a note, the styling is kept, which is good, but the newlines are removed.<br>
    Also when I share the selection to Notesnook via the share functionality from Android, I have the same issue.</p>
    <hr>
    <p dir="auto"><strong>Device information:</strong><br>
    App version: 2.3.0<br>
    Platform: android<br>
    Model: OnePlus-CPH2409-31<br>
    Pro: true<br>
    Logged in: yes</p><!--EndFragment-->
    </body>
    </html>) 1`] = `"<!--StartFragment--><p dir=\\"auto\\">When I try to paste something (e.g. email content) to a note, the styling is kept, which is good, but the newlines are removed.</p><p data-spacing=\\"single\\">    Also when I share the selection to Notesnook via the share functionality from Android, I have the same issue.</p>    <hr>    <p dir=\\"auto\\"><strong>Device information:</strong></p><p data-spacing=\\"single\\">    App version: 2.3.0</p><p data-spacing=\\"single\\">    Platform: android</p><p data-spacing=\\"single\\">    Model: OnePlus-CPH2409-31</p><p data-spacing=\\"single\\">    Pro: true</p><p data-spacing=\\"single\\">    Logged in: yes</p><!--EndFragment-->"`;

exports[`convert br tags to paragraphs (<html><body>
    <!--StartFragment--><span class="css-901oao css-16my406 r-poiln3 r-bcqeeo r-qvutc0">Why switch from Gmail? 

    Not sacrificing features for more privacy, prefer using one app, in many public groups and channels (Telegram)
    
    LibreOffice Slow &amp; buggy
    
    Switched to Brave for the better Android app, more private out of the box &amp; unsure if uBlock Origin closes gap</span><!--EndFragment-->
    </body>
    </html>) 1`] = `"<!--StartFragment--><span class=\\"css-901oao css-16my406 r-poiln3 r-bcqeeo r-qvutc0\\"><p>Why switch from Gmail? </p><p data-spacing=\\"single\\"></p><p data-spacing=\\"single\\">    Not sacrificing features for more privacy, prefer using one app, in many public groups and channels (Telegram)</p><p data-spacing=\\"single\\">    </p><p data-spacing=\\"single\\">    LibreOffice Slow &amp; buggy</p><p data-spacing=\\"single\\">    </p><p data-spacing=\\"single\\">    Switched to Brave for the better Android app, more private out of the box &amp; unsure if uBlock Origin closes gap</p></span><!--EndFragment-->"`;

exports[`convert br tags to paragraphs (<p><br/></p>) 1`] = `"<p><br></p>"`;

exports[`convert br tags to paragraphs (<p>line <em>1<br>line</em> 2</p>) 1`] = `"<p>line <em>1</em></p><p data-spacing=\\"single\\"><em>line</em> 2</p>"`;

exports[`convert br tags to paragraphs (<p>line <span><em>1<br data-some="hello">line</em></span> 2</p>) 1`] = `"<p>line <span><em>1</em></span></p><p data-spacing=\\"single\\"><span><em>line</em></span> 2</p>"`;

exports[`convert br tags to paragraphs (<p>line <span><em>1<br>line</em></span> 2</p>) 1`] = `"<p>line <span><em>1</em></span></p><p data-spacing=\\"single\\"><span><em>line</em></span> 2</p>"`;

exports[`convert br tags to paragraphs (<p>line 1<br>line 2</p>) 1`] = `"<p>line 1</p><p data-spacing=\\"single\\">line 2</p>"`;
